Add docs for Heat adoption

bshephar opened the pull request.

jistr commented on docs/openstack/heat_adoption.md:
```diff
@@ -0,0 +1,158 @@
+
+As already done for [Keystone](https://github.com/openstack-k8s-operators/data-plane-adoption/blob/main/keystone_adoption.md), the Heat Adoption follows a similar pattern.
+
+Patch the `osp-secret` to update the `HeatAuthEncryptionKey` password. This needs
```
I think both the Heat service password and auth encryption key should be added to:

https://github.com/openstack-k8s-operators/data-plane-adoption/blob/main/docs/openstack/backend_services_deployment.md

and to

https://github.com/openstack-k8s-operators/data-plane-adoption/blob/b4019f55eb29a8e6c8efaa8240a7cdf31de880cf/tests/roles/backend_services/tasks/main.yaml#L14

bshephar replied:
This was done, just didn't resolve the conversation. Resolving.

jistr commented on docs/openstack/heat_adoption.md:
````diff
@@ -0,0 +1,158 @@
+Patch OpenStackControlPlane to deploy Heat:
+
+```
+oc patch openstackcontrolplane openstack --type=merge --patch '
````
And this could have a corresponding new Ansible role `heat_adoption` in the test suite:

https://github.com/openstack-k8s-operators/data-plane-adoption/tree/main/tests/roles

included from both playbooks:

https://github.com/openstack-k8s-operators/data-plane-adoption/tree/main/tests/playbooks

The Standalone should now be deploying Heat, so it should be adoptable in CI i think.

https://github.com/openstack-k8s-operators/install_yamls/blob/c5e089ca59b3c15e28e9b65d968b238f573cfe00/devsetup/standalone/role.j2#L98-L101

bshephar replied:
Let me know if the latest revision to this PR addresses these now.

jistr replied:
Thanks, overall it looks good to me. I think before we fork into TP3 docs there may be opportunities to streamline the text a bit but i'd prefer getting something working merged over polishing details. (E.g. I'd just trust that the user set the `osp-secret` passwords correctly as the relevant document says. It's also a matter of consistency -- we don't tell the user to double check those secrets in any of the other service adoption documents.)

fao89 reviewed (commented):
please, add a link to this doc on:
https://github.com/openstack-k8s-operators/data-plane-adoption/blob/main/docs/README.md

https://github.com/openstack-k8s-operators/data-plane-adoption/blob/main/mkdocs.yml#L29

bshephar commented on docs/openstack/backend_services_deployment.md:
```diff
@@ -58,6 +58,8 @@ podified OpenStack control plane services.
   NOVA_PASSWORD=$(cat ~/tripleo-standalone-passwords.yaml | grep ' NovaPassword:' | awk -F ': ' '{ print $2; }')
   OCTAVIA_PASSWORD=$(cat ~/tripleo-standalone-passwords.yaml | grep ' OctaviaPassword:' | awk -F ': ' '{ print $2; }')
   PLACEMENT_PASSWORD=$(cat ~/tripleo-standalone-passwords.yaml | grep ' PlacementPassword:' | awk -F ': ' '{ print $2; }')
+  HEATAUTHENCRYPTIONKEY=$(cat ~/overcloud-deploy/overcloud/overcloud-passwords.yaml | grep ' HeatAuthEncryptionKey:' | awk -F ': ' '{ print $2; }')
+  HEATPASSWORD=$(cat ~/overcloud-deploy/overcloud/overcloud-passwords.yaml | grep ' HeatPassword:' | awk -F ': ' '{ print $2; }')
```
Eh, standalone path. I'll have to change this

bshephar commented:
> please, add a link to this doc on: https://github.com/openstack-k8s-operators/data-plane-adoption/blob/main/docs/README.md
> 
> https://github.com/openstack-k8s-operators/data-plane-adoption/blob/main/mkdocs.yml#L29

Ack, done.

cescgina commented on tests/playbooks/test_minimal.yaml:
```diff
@@ -34,3 +34,4 @@
     - cinder_adoption
     - horizon_adoption
     - dataplane_adoption
+    - heat_adoption
```
Currently in CI we use the `test_with_ceph.yaml` playbook instead of the `test_minimal.yaml`, so we should add the role there as well to test is. Logs for the run in this PR show that head adoption is not running: https://logserver.rdoproject.org/79/179/292f6f8f1d53f3ab9d8614461ddedebf48ea06b8/github-check/data-plane-adoption-github-rdo-centos-9-crc-single-node/ec4747e/controller/data-plane-adoption-tests-repo/data-plane-adoption/tests/logs/test_with_ceph_out_2023-10-13T01:38:24EDT.log

bshephar replied:
ack. Added now.

cescgina commented:
recheck

jistr commented on tests/secrets.sample.yaml:
```diff
@@ -19,6 +19,8 @@ cinder_password: "{{ lookup('file', tripleo_passwords) | from_yaml | community.g
 ironic_password: "{{ lookup('file', tripleo_passwords) | from_yaml | community.general.json_query('*.IronicPassword') | first }}"
 neutron_password: "{{ lookup('file', tripleo_passwords) | from_yaml | community.general.json_query('*.NeutronPassword') | first }}"
+heat_password: "{{ lookup('file', tripleo_passwords) | from_yaml | community.general.json_query('*.HeatPassword') | first }}"
+heat_auth_encryption_key: "{{ lookup('file', tripleo_passwords) | from_yaml | community.general.json_query('*.HeatAuthEncryptionKey') | first }}"
```
This may need wiring into CI to get it passing. Traditionally we've set each password separately there, but now that we have the default lookup-based method, we could utilize the `tripleo_passwords` file in CI too perhaps, and then we wouldn't have to wire in each new service separately. WDYT @marios @cescgina ?

Those can be separate concerns though. We can first wire in the heat passwords the old way and then switch everything to `tripleo_passwords` file if applicable.

cescgina replied:
sounds good to me, I crated https://review.rdoproject.org/r/c/rdo-jobs/+/50563 to unblock the merge and we can simplify the passwords handling in the future

cescgina replied:
Testing here: https://review.rdoproject.org/r/c/testproject/+/50564

jistr replied:
:+1: thanks

cescgina commented on tests/roles/backend_services/tasks/main.yaml:
```diff
@@ -37,6 +37,10 @@
     {% endif %}
+    {% if heat_password %}
+        oc set data secret/osp-secret "HeatPassword={{ heat_password }}"
+        oc set data secret/osp=secret "HeatAuthEncryptionKey={{ heat_auth_encryption_key }}"
```
there is a typo here, `oc set data secret/osp=secret "HeatAuthEncryptionKey={{ heat_auth_encryption_key }}"` should be `oc set data secret/osp-secret "HeatAuthEncryptionKey={{ heat_auth_encryption_key }}"`

bshephar replied:
Ah, fixed. Thanks for pointing that out.

cescgina commented:
In last run it looks like Heat was not deployed as it was waiting for transportURL: https://logserver.rdoproject.org/79/179/b019932e69515071d1f0c4f87d4ac6626bad5840/github-check/data-plane-adoption-github-rdo-centos-9-extracted-crc/37fdf13/controller/crs/heats.yaml and the TransportURL itself was started but not finished: https://logserver.rdoproject.org/79/179/b019932e69515071d1f0c4f87d4ac6626bad5840/github-check/data-plane-adoption-github-rdo-centos-9-extracted-crc/37fdf13/controller/crs/transporturls.yaml

bshephar commented:
> https://logserver.rdoproject.org/79/179/b019932e69515071d1f0c4f87d4ac6626bad5840/github-check/data-plane-adoption-github-rdo-centos-9-extracted-crc/37fdf13/controller/crs/transporturls.yaml

Yeah, I don't see much in the rabbitmq operator logs to suggest why it was a problem. Maybe just a timing thing? Let's recheck and see if the error reproduces.

bshephar commented:
recheck

bshephar commented:
Right, so in my deployment a lot of the fields are getting pre-filled:
```
❯ oc get openstackcontrolplane openstack-galera -o yaml | yq .spec.heat
apiOverride: {}
cnfAPIOverride: {}
enabled: true
template:
  databaseInstance: openstack
  databaseUser: heat
  debug:
    dbSync: false
  heatAPI:
    containerImage: ""
    debug:
      service: false
    override: {}
    replicas: 1
    resources: {}
  heatCfnAPI:
    containerImage: ""
    debug:
      service: false
    override: {}
    replicas: 1
    resources: {}
  heatEngine:
    containerImage: ""
    debug:
      service: false
    replicas: 1
    resources: {}
  memcachedInstance: memcached
  passwordSelectors:
    authEncryptionKey: HeatAuthEncryptionKey
    database: HeatDatabasePassword
    service: HeatPassword
  preserveJobs: false
  rabbitMqClusterName: rabbitmq
  secret: osp-secret
  serviceUser: heat
```

But checking the logs from this CI environment:
```
❯ curl -s https://logserver.rdoproject.org/79/179/b019932e69515071d1f0c4f87d4ac6626bad5840/github-check/data-plane-adoption-github-rdo-centos-9-crc-single-node/e880464/controller/crs/openstackcontrolplanes.yaml | yq ".items[].spec.heat"
apiOverride:
  route: {}
cnfAPIOverride: {}
enabled: true
template:
  databaseInstance: openstack
  databaseUser: ""
  debug:
    dbSync: false
  heatAPI:
    containerImage: ""
    debug:
      service: false
    override: {}
    replicas: 1
    resources: {}
  heatCfnAPI:
    containerImage: ""
    debug:
      service: false
    override: {}
    replicas: 1
    resources: {}
  heatEngine:
    containerImage: ""
    debug:
      service: false
    replicas: 1
    resources: {}
  memcachedInstance: memcached
  passwordSelectors:
    authEncryptionKey: HeatAuthEncryptionKey
    database: HeatDatabasePassword
    service: HeatPassword
  preserveJobs: false
  rabbitMqClusterName: ""
  secret: osp-secret
  serviceUser: ""
```

So, the issue seems to be that we're not populating those fields, in particular `rabbitMqClusterName` isn't populated here. We also need `databaseUser` and `serviceUser`.

bshephar commented:
Didn't make it far enough to configure heat this time:
```
❯ curl -s https://logserver.rdoproject.org/79/179/e1f6a558e38bbcb102807a3956c0470908b8e952/github-check/data-plane-adoption-github-rdo-centos-9-extracted-crc/3bced2f/controller/crs/openstackcontrolplanes.yaml | yq '.items[].spec.heat.enabled'
false
```

It failed to configure Placement. Which I don't think will be related to this change. Rechecking again to see if it reproduces.

bshephar commented:
recheck

bshephar commented on tests/roles/heat_adoption/tasks/main.yaml:
```diff
@@ -0,0 +1,56 @@
+            [DEFAULT]
+            debug = true
+            [keystone_authtoken]
+            insecure = true
```
Looks like something might have changed recently and Heat now fails to connect to Keystone because of the self-signed cert. I'll need to find the CA and have the Heat containers trust the CA. Adding this here to see if the CI jobs will pass in the meantime, or find any additional issues that might exist.

bshephar commented:
Ok, latest `crc-single-node` failed on the step after Heat becomes ready:
```
TASK [heat_adoption : check that heat is reachable and its endpoints are defined] ***
changed: [localhost] => {"attempts": 1, "changed": true, "cmd": "set -euxo pipefail\n\n\nalias openstack=\"oc exec -t openstackclient -- openstack\"\n\n${BASH_ALIASES[openstack]} endpoint list -c \"Service Name\" -f yaml | grep heat\n${BASH_ALIASES[openstack]} stack list\n", "delta": "0:00:28.303535", "end": "2023-11-02 00:48:41.563050", "msg": "", "rc": 0, "start": "2023-11-02 00:48:13.259515", "stderr": "+ alias 'openstack=oc exec -t openstackclient -- openstack'\n+ oc exec -t openstackclient -- openstack endpoint list -c 'Service Name' -f yaml\n+ grep heat\n+ oc exec -t openstackclient -- openstack stack list", "stderr_lines": ["+ alias 'openstack=oc exec -t openstackclient -- openstack'", "+ oc exec -t openstackclient -- openstack endpoint list -c 'Service Name' -f yaml", "+ grep heat", "+ oc exec -t openstackclient -- openstack stack list"], "stdout": "- Service Name: heat\n- Service Name: heat-cfn\n- Service Name: heat-cfn\n- Service Name: heat", "stdout_lines": ["- Service Name: heat", "- Service Name: heat-cfn", "- Service Name: heat-cfn", "- Service Name: heat"]}

TASK [dataplane_adoption : force latest ansible-runner image] ******************
fatal: [localhost]: FAILED! => {"changed": true, "cmd": "set -euxo pipefail\n\n\noc login -u kubeadmi
```

Extracted job failed to start CRC. So we'll recheck to see if that one passes now.

bshephar commented:
recheck

cescgina reviewed (approved):
lgtm, tests passed in extracted crc job and the single-node failure is unrelated
